### PR TITLE
feature#1

### PR DESCRIPTION
**Pull request by seungheondoh**
# Pull Request
레파지토리에 기여해주셔서 감사드립니다.

해당 PR을 제출하기 전에 아래 사항이 완료되었는지 확인 부탁드립니다:
- [x] 작성한 코드가 어떤 에러나 경고없이 빌드가 되었나요?
- [x] 충분한 테스트를 수행하셨나요?

## 1. 해당 PR은 어떤 내용인가요?
모델의 핵심 아키텍쳐 내용입니다. one_batch에 대해서 forward pass를 확인했습니다.

### CODE REVIEW

File: src/model/net.py [hunk 1]
```diff
@@ -1,4 +1,86 @@
...
+        sample_rate=16000,
+        n_fft=512,
+        f_min=0.0,
+        f_max=8000.0,
+        n_mels=128,
+        n_class=50,
+        n_harmonic=6,
+        semitone_scale=2,
+        learn_bw="only_Q",
+    ):
```
Comment (ssaru) [inline src/model/net.py]: typing hinting 해주시면 좋을 것 같습니다.

File: src/model/ops.py [hunk 2]
```diff
@@ -1,16 +1,317 @@
...
+        self.f0 = torch.tensor(harmonic_hz.astype("float32"))
+
+        # Bandwidth parameters
+        if learn_bw == "only_Q":
+            self.bw_Q = nn.Parameter(torch.tensor(np.array([bw_Q]).astype("float32")))
+        elif learn_bw == "fix":
+            self.bw_Q = torch.tensor(np.array([bw_Q]).astype("float32"))
+
+    def get_harmonic_fb(self):
+        # bandwidth
```
Comment (ssaru) [inline src/model/ops.py]: 함수 docstring을 적어주시면 좋을 것 같아요.

File: src/model/ops.py [hunk 3]
```diff
@@ -1,16 +1,317 @@
...
+        harmonic_fb = self.get_harmonic_fb()
+        harmonic_spec = torch.matmul(
+            spectrogram.transpose(1, 2), harmonic_fb
+        ).transpose(1, 2)
+
+        # (batch, channel, length) -> (batch, harmonic, f0, length)
+        b, c, l = harmonic_spec.size()
+        harmonic_spec = harmonic_spec.view(b, self.n_harmonic, self.level, l)
+
+        # amplitude to db
```
Comment (ssaru) [inline src/model/ops.py]: 이미 밑에 코드에 `self.amplitude_to_db(harmonic_spec)`가 있기 때문에
주석으로써의 `# amplitude to db`는 의미없는 주석같습니다.

```suggestion
```

**Review by dreamgonfly — APPROVED** [review]

File: src/model/net.py [hunk 4]
```diff
@@ -1,4 +1,86 @@
...
+            x = nn.MaxPool1d(x.size(-1))(x)
+        x = x.squeeze(2)
+
+        # Dense
+        x = self.dense1(x)
+        x = self.bn(x)
+        x = self.relu(x)
+        x = self.dropout(x)
+        x = self.dense2(x)
+        x = nn.Sigmoid()(x)
```
Comment (seopbo) [inline src/model/net.py]: `nn.Sigmoid`를 함수형으로 쓰는 게 좋을 것 같습니다. 이러면 `foward`할 때마다 instance가 생깁니다.

File: src/model/ops.py [hunk 5]
```diff
@@ -1,16 +1,317 @@
...
-def add(a, b):
-    return a + b
+    def forward(self, x):
+        out = self.mp(self.relu(self.bn(self.conv(x))))
+        return out
 
 
-def subtract(a, b):
-    return a - b
+class Res_2d_mp(nn.Module):
```
Comment (ssaru) [inline src/model/ops.py]: python의 클래스 컨벤션은 CamelCase로 쓰는게 좋을 것 같습니다.

```suggestion
class Res2Dmp(nn.Module):
```

File: src/model/net.py [hunk 6]
```diff
@@ -1,4 +1,86 @@
...
+        x = self.layer1(x)
+        x = self.layer2(x)
+        x = self.layer3(x)
+        x = self.layer4(x)
+        x = self.layer5(x)
+        x = self.layer6(x)
+        x = self.layer7(x)
+        x = x.squeeze(2)
+
+        # Global Max Pooling
```
Comment (seopbo) [inline src/model/net.py:83]: `nn.Maxpool1d`를 함수형으로 쓰는 게 좋을 것 같습니다. 이러면 `foward`할 때마다 instance가 생깁니다.

File: src/model/ops.py [hunk 7]
```diff
@@ -1,16 +1,317 @@
...
+        n_fft=513,
+        win_length=None,
+        hop_length=None,
+        pad=0,
+        power=2,
+        normalized=False,
+        n_harmonic=6,
+        semitone_scale=2,
+        bw_Q=1.0,
+        learn_bw=None,
```
Comment (ssaru) [inline src/model/ops.py]: type hint를 주시면 좋을 것 같습니다.

File: src/model/ops.py [hunk 8]
```diff
@@ -1,16 +1,317 @@
...
+        bw = bw.unsqueeze(0)  # (1, n_band)
+        f0 = self.f0.unsqueeze(0)  # (1, n_band)
+        fft_bins = self.fft_bins.unsqueeze(1)  # (n_bins, 1)
+
+        up_slope = torch.matmul(fft_bins, (2 / bw)) + 1 - (2 * f0 / bw)
+        down_slope = torch.matmul(fft_bins, (-2 / bw)) + 1 + (2 * f0 / bw)
+        fb = torch.max(self.zero, torch.min(down_slope, up_slope))
+        return fb
+
+    def to_device(self, device, n_bins):
```
Comment (ssaru) [inline src/model/ops.py]: type hint주시면 좋을 것 같습니다.

File: src/model/ops.py [hunk 9]
```diff
@@ -1,16 +1,317 @@
...
+
+        # amplitude to db
+        harmonic_spec = self.amplitude_to_db(harmonic_spec)
+        return harmonic_spec
+
+
+# Modules for harmonic filters
+
+
+def initialize_filterbank(sample_rate, n_harmonic, semitone_scale):
```
Comment (ssaru) [inline src/model/ops.py]: 여기도 type hint요!

File: src/model/ops.py [hunk 10]
```diff
@@ -1,16 +1,317 @@
...
+    # MIDI
+    # lowest note
```
Comment (ssaru) [inline src/model/ops.py]: lowest note는 어떤 의미인가요?

File: src/model/ops.py [hunk 11]
```diff
@@ -1,16 +1,317 @@
...
+        "#": 1,
+        "": 0,
+        "b": -1,
+        "!": -1,
+        "♯": 1,
+        "𝄪": 2,
+        "♭": -1,
+        "𝄫": -2,
+        "♮": 0,
+    }
```
Comment (ssaru) [inline src/model/ops.py]: 상수변수로 빼면 더 좋을 것 같습니다.

File: src/model/ops.py [hunk 12]
```diff
@@ -1,16 +1,317 @@
...
+    match = re.match(
+        r"^(?P<tonic>[A-Ga-g])"
+        r"(?P<accidental>[#♯b!♭]?)"
+        r":(?P<scale>(maj|min)(or)?)$",
+        key,
+    )
+    if not match:
+        raise ParameterError("Improper key format: {:s}".format(key))
+    pitch_map = {"C": 0, "D": 2, "E": 4, "F": 5, "G": 7, "A": 9, "B": 11}
+    acc_map = {"#": 1, "": 0, "b": -1, "!": -1, "♯": 1, "♭": -1}
```
Comment (ssaru) [inline src/model/ops.py]: 상수변수로 빼면 더 좋을 것 같습니다.

File: src/model/ops.py [hunk 13]
```diff
@@ -1,16 +1,317 @@
...
+
+    elif 0 <= tonic_number < 6:
+        use_sharps = True
+
+    elif tonic_number > 6:
+        use_sharps = False
+
+    # Basic note sequences for simple keys
+    notes_sharp = ["C", "C♯", "D", "D♯", "E", "F", "F♯", "G", "G♯", "A", "A♯", "B"]
+    notes_flat = ["C", "D♭", "D", "E♭", "E", "F", "G♭", "G", "A♭", "A", "B♭", "B"]
```
Comment (ssaru) [inline src/model/ops.py]: 상수변수로 빼면 더 좋을 것 같습니다.

File: src/model/ops.py [hunk 14]
```diff
@@ -1,16 +1,317 @@
...
+    ]
+
+    flat_corrections = [
+        (11, "C♭"),
+        (4, "F♭"),
+        (9, "B𝄫"),
+        (2, "E𝄫"),
+        (7, "A𝄫"),
+        (0, "D𝄫"),
+    ]  # last would be (5, 'G𝄫')
```
Comment (ssaru) [inline src/model/ops.py]: 상수변수로 빼면 더 좋을 것 같습니다.
